File: static/cpp/java/src/main/java/com/epi/SmallestSubarrayCoveringSetStream.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import static com.epi.utils.Utils.objectInputStreamFromList;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class SmallestSubarrayCoveringSetStream {
  // @include
  // Represent subarray by starting and ending indices, inclusive.
  private static class Subarray {
    public Integer start;
    public Integer end;

    public Subarray(Integer start, Integer end) {
      this.start = start;
      this.end = end;
    }
  }

  private static Integer getValueForFirstEntry(
      LinkedHashMap<String, Integer> m) {
    // LinkedHashMap guarantees iteration over key-value pairs takes place in
    // insertion order, most recent first.
    Integer result = null;
    for (Map.Entry<String, Integer> entry : m.entrySet()) {
      result = entry.getValue();
      break;
    }
    return result;
  }

  public static Subarray findSmallestSubarrayCoveringSubset(
      Iterator<String> iter, List<String> queryStrings) {
    LinkedHashMap<String, Integer> dict = new LinkedHashMap<>();
    for (String s : queryStrings) {
      dict.put(s, null);
    }
    int numStringsFromQueryStringsSeenSoFar = 0;

    Subarray res = new Subarray(-1, -1);
    int idx = 0;
    while (iter.hasNext()) {
      String s = iter.next();
      if (dict.containsKey(s)) { // s is in queryStrings.
        Integer it = dict.get(s);
        if (it == null) {
          // First time seeing this string from queryStrings.
          numStringsFromQueryStringsSeenSoFar++;
        }
        // dict.put(s,idx) won't work because it does not move the entry to
        // the front of the queue if an entry with key s is already present.
        // So we explicitly remove the existing entry with key s, then put
        // (s,idx).
        dict.remove(s);
        dict.put(s, idx);
      }

      if (numStringsFromQueryStringsSeenSoFar == queryStrings.size()) {
        // We have seen all strings in queryStrings, let's get to work.
        if ((res.start == -1 && res.end == -1)
            || idx - getValueForFirstEntry(dict) < res.end - res.start) {
          res.start = getValueForFirstEntry(dict);
          res.end = idx;
        }
      }
      ++idx;
    }
    return res;
  }
  // @exclude

  public static Integer[] findSmallestSubarrayCoveringSubsetAsArray(
      List<String> A, List<String> queryStrings) {
    Subarray result
        = findSmallestSubarrayCoveringSubset(A.iterator(), queryStrings);
    Integer[] resultArray = new Integer[2];
    resultArray[0] = result.start;
    resultArray[1] = result.end;
    return resultArray;
  }
}

File: solutions/python/smallest_subarray_covering_set_stream.py
# @include
def find_smallest_subarray_covering_subset(stream, query_strings):
    class DoublyLinkedListNode:

        def __init__(self, data=None):
            self.data = data
            self.next = self.prev = None

    class LinkedList:

        def __init__(self):
            self.head = self.tail = None
            self._size = 0

        def __len__(self):
            return self._size

        def insert_after(self, value):
            node = DoublyLinkedListNode(value)
            node.prev = self.tail
            if self.tail:
                self.tail.next = node
            else:
                self.head = node
            self.tail = node
            self._size += 1

        def remove(self, node):
            if node.next:
                node.next.prev = node.prev
            else:
                self.tail = node.prev
            if node.prev:
                node.prev.next = node.next
            else:
                self.head = node.next
            node.next = node.prev = None
            self._size -= 1

    # Tracks the last occurrence (index) of each string in query_strings.
    loc = LinkedList()
    d = {s: None for s in query_strings}
    res = (-1, -1)
    for idx, s in enumerate(stream):
        if s in d:  # s is in query_strings.
            it = d[s]
            if it is not None:
                # Explicitly remove s so that when we add it, it's the string most
                # recently added to loc.
                loc.remove(it)
            loc.insert_after(idx)
            d[s] = loc.tail

            if len(loc) == len(query_strings):
                # We have seen all strings in query_strings, let's get to work.
                if res == (-1, -1) or idx - loc.head.data < res[1] - res[0]:
                    res = (loc.head.data, idx)
    return res
# @exclude
